### PR TITLE
Add case insensitive prefix and wildcard to 'version' field

### PR DESCRIPTION
**Pull request by cbuescher**
This change adds support for the recently introduced case insensitivity flag for
wildcard and prefix queries. Since version field values are encoded differently we
need to adapt our own AutomatonQuery variation to add both cases if case insensitivity
is turned on.

### CODE REVIEW

**Comment by cbuescher** [comment]
@markharwood would you mind taking a look since you worked on the case insensitivity option lately?

File: server/src/main/java/org/elasticsearch/common/lucene/search/AutomatonQueries.java [hunk 1]
```diff
@@ -129,11 +129,11 @@ public static Automaton toCaseInsensitiveString(String s, int maxDeterminizedSta
...
         a = MinimizationOperations.minimize(a, maxDeterminizedStates);
         return a;
- 
-    
+
+
     }
 
-    protected static Automaton toCaseInsensitiveChar(int codepoint, int maxDeterminizedStates) {
+    public static Automaton toCaseInsensitiveChar(int codepoint, int maxDeterminizedStates) {
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/common/lucene/search/AutomatonQueries.java:136]: Wanted to use this, let me know if this shouldn't be that open. Another option is simpyl copying the logic over I think.

Comment (markharwood) [reply server/src/main/java/org/elasticsearch/common/lucene/search/AutomatonQueries.java:136]: LGTM

**Comment by markharwood** [comment]
I'm working on backporting the case insensitive term/prefix/wildcard query support to 7.x [here](https://github.com/elastic/elasticsearch/pull/62661)
If it's OK with you I suggest that I complete that backport without adding any specialised CI support for the Version field and then you can backport this to fill in that hole?

File: x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java [hunk 2]
```diff
@@ -139,7 +139,7 @@ public Query existsQuery(QueryShardContext context) {
 
         @Override
         public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, boolean caseInsensitve, QueryShardContext context) {
-            return wildcardQuery(value + "*", method, context);
+            return wildcardQuery(value + "*", method, caseInsensitve, context);
         }
```
Comment (markharwood) [inline x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:143]: Typo - missing i in Insensitive.

File: x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java [hunk 3]
```diff
@@ -244,7 +244,7 @@ public Query wildcardQuery(String value, MultiTermQuery.RewriteMethod method, bo
                 );
             }
 
-            VersionFieldWildcardQuery query = new VersionFieldWildcardQuery(new Term(name(), value));
+            VersionFieldWildcardQuery query = new VersionFieldWildcardQuery(new Term(name(), value), caseInsensitve);
             QueryParsers.setRewriteMethod(query, method);
```
Comment (markharwood) [inline x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java:248]: Typo

**Review by markharwood — APPROVED** [review]
Couple of typos but otherwise LGTM

**Comment by cbuescher** [comment]
@markharwood thanks for the review. Sure I'll wait for your backport to 7.x before backporting this PR then.

**Comment by markharwood** [comment]
@cbuescher The 7.x backport is merged now

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Comment by cbuescher** [comment]
@elasticmachine update branch

**Pull request merged**